File: sami/step_definitions/Save Payment Information.js
import { Given, When, Then } from 'cypress-cucumber-preprocessor/steps';
import PaymentInfoPage from '../page-objects/PaymentInfoPage';

const paymentInfoPage = new PaymentInfoPage();

Given('the user is logged into their account', () => {
  // Code to ensure the user is logged in
});

Given('the user is on the {string} page', (page) => {
  paymentInfoPage.navigateToPage(page);
});

Given('the user enters valid payment details', (dataTable) => {
  const data = dataTable.rowsHash();
  paymentInfoPage.enterPaymentDetails(data);
});

Given('the user enters invalid payment details', (dataTable) => {
  const data = dataTable.rowsHash();
  paymentInfoPage.enterPaymentDetails(data);
});

Given('the user enters incomplete payment details', (dataTable) => {
  const data = dataTable.rowsHash();
  paymentInfoPage.enterPaymentDetails(data);
});

Given('the user has previously saved payment information', () => {
  paymentInfoPage.checkSavedPaymentInfo();
});

When('the user saves the payment information', () => {
  paymentInfoPage.savePaymentInfo();
});

When('the user attempts to save the payment information', () => {
  paymentInfoPage.savePaymentInfo();
});

When('the user updates the payment details', (dataTable) => {
  const data = dataTable.rowsHash();
  paymentInfoPage.enterPaymentDetails(data);
});

When('the user saves the updated payment information', () => {
  paymentInfoPage.savePaymentInfo();
});

When('the user opts to delete the saved payment information', () => {
  paymentInfoPage.deletePaymentInfo();
});

Then('the payment information should be saved securely', () => {
  paymentInfoPage.verifyPaymentInfoSaved();
});

Then('a confirmation message {string} should be displayed', (message) => {
  paymentInfoPage.verifyConfirmationMessage(message);
});

Then('an error message {string} should be displayed', (message) => {
  paymentInfoPage.verifyErrorMessage(message);
});

Then('the payment information should not be saved', () => {
  paymentInfoPage.verifyPaymentInfoNotSaved();
});

Then('the updated payment information should be saved securely', () => {
  paymentInfoPage.verifyPaymentInfoUpdated();
});

Then('the payment information should be removed', () => {
  paymentInfoPage.verifyPaymentInfoDeleted();
});

File: step_definitions/Save Payment Information.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class PaymentInformationSteps extends PaymentInformationPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is logged into their account")
    public void theUserIsLoggedIntoTheirAccount() {
        loginToAccount();
    }

    @Given("the user is on the checkout page")
    public void theUserIsOnTheCheckoutPage() {
        navigateToCheckoutPage();
    }

    @When("the user enters valid payment information")
    public void theUserEntersValidPaymentInformation() {
        String cardNumber = testHarness.getData("PaymentData", "CardNumber");
        String expiryDate = testHarness.getData("PaymentData", "ExpiryDate");
        String cvv = testHarness.getData("PaymentData", "CVV");
        enterPaymentInformation(cardNumber, expiryDate, cvv);
    }

    @When("the user chooses to save the payment information for future purchases")
    public void theUserChoosesToSaveThePaymentInformationForFuturePurchases() {
        chooseToSavePaymentInformation(true);
    }

    @When("the user chooses not to save the payment information")
    public void theUserChoosesNotToSaveThePaymentInformation() {
        chooseToSavePaymentInformation(false);
    }

    @When("the user submits the payment information")
    public void theUserSubmitsThePaymentInformation() {
        submitPaymentInformation();
    }

    @Then("the payment information is saved securely")
    public void thePaymentInformationIsSavedSecurely() {
        verifyPaymentInformationSaved();
    }

    @Then("a confirmation message is displayed to the user")
    public void aConfirmationMessageIsDisplayedToTheUser() {
        verifyConfirmationMessageDisplayed();
    }

    @Then("the payment information is not saved")
    public void thePaymentInformationIsNotSaved() {
        verifyPaymentInformationNotSaved();
    }

    @Then("the transaction is completed successfully")
    public void theTransactionIsCompletedSuccessfully() {
        verifyTransactionCompleted();
    }

    @When("the user enters invalid payment information")
    public void theUserEntersInvalidPaymentInformation() {
        String cardNumber = testHarness.getData("PaymentData", "InvalidCardNumber");
        String expiryDate = testHarness.getData("PaymentData", "InvalidExpiryDate");
        String cvv = testHarness.getData("PaymentData", "InvalidCVV");
        enterPaymentInformation(cardNumber, expiryDate, cvv);
    }

    @Then("an error message is displayed indicating the payment information is invalid")
    public void anErrorMessageIsDisplayedIndicatingThePaymentInformationIsInvalid() {
        verifyErrorMessageDisplayed();
    }

    @Given("the user has previously saved payment information")
    public void theUserHasPreviouslySavedPaymentInformation() {
        verifySavedPaymentInformationExists();
    }

    @When("the user navigates to the payment information section of their account")
    public void theUserNavigatesToThePaymentInformationSectionOfTheirAccount() {
        navigateToPaymentInformationSection();
    }

    @Then("the saved payment information is displayed securely")
    public void theSavedPaymentInformationIsDisplayedSecurely() {
        verifySavedPaymentInformationDisplayed();
    }

    @Then("the user can edit or delete the payment information")
    public void theUserCanEditOrDeleteThePaymentInformation() {
        verifyEditOrDeleteOptionsAvailable();
    }

    @When("the user chooses to delete the saved payment information")
    public void theUserChoosesToDeleteTheSavedPaymentInformation() {
        deleteSavedPaymentInformation();
    }

    @Then("the payment information is removed from the account")
    public void thePaymentInformationIsRemovedFromTheAccount() {
        verifyPaymentInformationRemoved();
    }

    @When("the user edits the saved payment information")
    public void theUserEditsTheSavedPaymentInformation() {
        String newCardNumber = testHarness.getData("PaymentData", "NewCardNumber");
        String newExpiryDate = testHarness.getData("PaymentData", "NewExpiryDate");
        String newCvv = testHarness.getData("PaymentData", "NewCVV");
        editPaymentInformation(newCardNumber, newExpiryDate, newCvv);
    }

    @When("the user submits the updated payment information")
    public void theUserSubmitsTheUpdatedPaymentInformation() {
        submitUpdatedPaymentInformation();
    }

    @Then("the payment information is updated securely")
    public void thePaymentInformationIsUpdatedSecurely() {
        verifyPaymentInformationUpdated();
    }
}